File: src/python/xerxes/mcp/reconnect.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Exponential-backoff reconnect for flaky MCP connections.

MCP servers are subprocesses or remote endpoints that occasionally drop.
:func:`reconnect_with_backoff` wraps an arbitrary connect callable in a
retry loop (5 attempts at 1s, 2s, 4s, 8s, 16s by default).
Exception messages are passed through :func:`scrub_credentials` so leaked
API keys / bearer tokens don't end up in logs or the TUI.
"""

from __future__ import annotations

import re
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import TypeVar

T = TypeVar("T")


class MCPReconnectError(RuntimeError):
    """Raised when :func:`reconnect_with_backoff` exhausts all attempts.

    Wraps the final underlying failure with a single, credential-scrubbed
    message. The original exception is preserved via ``__cause__`` (``raise
    ... from``). Using a dedicated single-argument exception avoids trying to
    reconstruct the original type, which fails for exceptions whose
    constructor is not ``(str) -> exc`` (e.g. ``ExceptionGroup`` /
    ``BaseExceptionGroup`` raised by anyio task groups in the MCP SDK
    transports).
    """


@dataclass
class ReconnectPolicy:
    """Exponential-backoff parameters for :func:`reconnect_with_backoff`.

    Attributes:
        max_attempts: Total attempts before giving up.
        base_seconds: Delay before the second attempt.
        factor: Multiplier applied after every failed attempt.
        max_seconds: Upper bound on any single delay.
    """

    max_attempts: int = 5
    base_seconds: float = 1.0
    factor: float = 2.0
    max_seconds: float = 60.0


_CRED_PATTERNS = [
    re.compile(r"(api[_-]?key)[\s:=\"']+([A-Za-z0-9._\-]{8,})", re.I),
    re.compile(r"(token)[\s:=\"']+([A-Za-z0-9._\-]{16,})", re.I),
    re.compile(r"(authorization:\s*bearer)\s+([A-Za-z0-9._\-]+)", re.I),
    re.compile(r"(password)[\s:=\"']+(\S+)", re.I),
    re.compile(r"sk-[A-Za-z0-9_\-]{16,}"),
]


def scrub_credentials(text: str) -> str:
    """Replace api-key / bearer-token / password fragments in ``text`` with ``[redacted]``."""
    out = text
    for pattern in _CRED_PATTERNS:
        # Simple ``re.sub`` with a function that preserves the label half.
        out = pattern.sub(lambda m: f"{m.group(1)}=[redacted]" if m.groups() else "[redacted]", out)
    return out


def reconnect_with_backoff(
    connect: Callable[[], T],
    *,
    policy: ReconnectPolicy | None = None,
    sleep: Callable[[float], None] = time.sleep,
    on_error: Callable[[int, BaseException], None] | None = None,
) -> T:
    """Retry ``connect`` with exponential backoff until success or ``max_attempts``.

    Each failure invokes ``on_error(attempt, exc)`` if supplied; after the
    last attempt fails the most recent exception is re-raised with its
    message scrubbed of credentials. ``sleep`` is injectable so tests can
    skip real time.
    """

    p = policy or ReconnectPolicy()
    last_exc: BaseException | None = None
    for attempt in range(1, p.max_attempts + 1):
        try:
            return connect()
        except BaseException as exc:
            last_exc = exc
            if on_error is not None:
                on_error(attempt, exc)
            if attempt >= p.max_attempts:
                break
            delay = min(p.max_seconds, p.base_seconds * (p.factor ** (attempt - 1)))
            sleep(delay)
    # Re-raise with credentials scrubbed. Do not reconstruct the original
    # exception type -- many exceptions (e.g. ExceptionGroup) cannot be
    # rebuilt from a single string. Wrap in a dedicated single-message error
    # and chain the original via ``from`` so it is never lost.
    assert last_exc is not None
    scrubbed = scrub_credentials(str(last_exc))
    raise MCPReconnectError(scrubbed) from last_exc


__all__ = ["MCPReconnectError", "ReconnectPolicy", "reconnect_with_backoff", "scrub_credentials"]

File: src/typescript/src/mcp/reconnect.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

/** Parameters for the MCP connection retry schedule. Delays are measured in seconds. */
export interface ReconnectPolicyOptions {
  readonly baseSeconds?: number
  readonly factor?: number
  readonly maxAttempts?: number
  readonly maxSeconds?: number
}

/** Exponential-backoff policy used when an MCP server disconnects. */
export class ReconnectPolicy {
  readonly baseSeconds: number
  readonly factor: number
  readonly maxAttempts: number
  readonly maxSeconds: number

  constructor(options: ReconnectPolicyOptions = {}) {
    this.maxAttempts = positiveInteger(options.maxAttempts ?? 5, 'maxAttempts')
    this.baseSeconds = positiveNumber(options.baseSeconds ?? 1, 'baseSeconds')
    this.factor = positiveNumber(options.factor ?? 2, 'factor')
    this.maxSeconds = positiveNumber(options.maxSeconds ?? 60, 'maxSeconds')
  }

  /** Return the delay after a one-based failed connection attempt. */
  delayForAttempt(attempt: number): number {
    const normalizedAttempt = Math.max(1, Math.floor(attempt))
    return Math.min(this.maxSeconds, this.baseSeconds * this.factor ** (normalizedAttempt - 1))
  }
}

/** Raised after the final MCP connection retry fails. Its message is always credential-scrubbed. */
export class MCPReconnectError extends Error {
  readonly attempts: number

  constructor(message: string, attempts: number) {
    super(scrubCredentials(message))
    this.name = new.target.name
    this.attempts = attempts
  }
}

export interface ReconnectWithBackoffOptions {
  /** Called after each failed attempt, before the next delay. */
  readonly onError?: (attempt: number, error: unknown) => void | Promise<void>
  readonly policy?: ReconnectPolicy | ReconnectPolicyOptions
  /** Injectable sleep implementation. It receives seconds, not milliseconds. */
  readonly sleep?: (seconds: number) => void | Promise<void>
}

/**
 * Retry a connection operation with exponential backoff.
 *
 * The retry hook receives the original error so hosts can apply their own
 * observability policy. The thrown terminal error only exposes a redacted
 * message so credentials from subprocess or HTTP diagnostics cannot escape
 * through the normal lifecycle API.
 */
export async function reconnectWithBackoff<T>(
  connect: () => T | Promise<T>,
  options: ReconnectWithBackoffOptions = {},
): Promise<T> {
  const policy = options.policy instanceof ReconnectPolicy
    ? options.policy
    : new ReconnectPolicy(options.policy)
  const sleep = options.sleep ?? sleepSeconds
  let lastError: unknown

  for (let attempt = 1; attempt <= policy.maxAttempts; attempt += 1) {
    try {
      return await connect()
    } catch (error) {
      lastError = error
      await options.onError?.(attempt, error)
      if (attempt >= policy.maxAttempts) {
        break
      }
      await sleep(policy.delayForAttempt(attempt))
    }
  }

  throw new MCPReconnectError(errorMessage(lastError), policy.maxAttempts)
}

/** Replace common API key, bearer token, and password fragments with a safe marker. */
export function scrubCredentials(text: string): string {
  return text
    .replace(/\b(password)\b\s*(?::|=|\s)\s*['"]?([^\s'";,]+)/gi, '$1=[redacted]')
    .replace(/\b(api[_-]?key)\b\s*(?::|=|\s)\s*['"]?([A-Za-z0-9._-]{8,})/gi, '$1=[redacted]')
    .replace(/\b(token)\b\s*(?::|=|\s)\s*['"]?([A-Za-z0-9._-]{16,})/gi, '$1=[redacted]')
    .replace(/\b(authorization\s*:\s*bearer)\s+([A-Za-z0-9._-]+)/gi, '$1=[redacted]')
    .replace(/\bsk-[A-Za-z0-9_-]{16,}\b/g, '[redacted]')
}

function errorMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message
  }
  return String(error)
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 1) {
    throw new RangeError(name + ' must be an integer of at least 1')
  }
  return value
}

function positiveNumber(value: number, name: string): number {
  if (!Number.isFinite(value) || value <= 0) {
    throw new RangeError(name + ' must be positive')
  }
  return value
}

function sleepSeconds(seconds: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, seconds * 1_000))
}
